iocore/eventsystem/I_Continuation.h

```diff
@@ -63,6 +63,35 @@ extern EThread *this_event_thread();
 
 typedef int (Continuation::*ContinuationHandler)(int event, void *data);
 
+// Convert event handler pointer fp to type ContinuationHandler, but with a compiler error if class C is not
+// derived from the class Continuation.
+//
+template <class C, typename T>
+constexpr ContinuationHandler
+continuation_handler_void_ptr(int (C::*fp)(int, T *))
+{
+  auto fp2 = reinterpret_cast<int (C::*)(int, void *)>(fp);
+  return static_cast<ContinuationHandler>(fp2);
+}
+
+// Overload for nullptr.
+//
+constexpr ContinuationHandler continuation_handler_void_ptr(std::nullptr_t)
+{
+#undef X
+#if !defined(__GNUC__)
+#define X 1
+#else
+#define X (__GNUC__ > 7)
```

Comment:
We can use `&&` like below?
```
#if defined(__GNUC__) && (__GNUC__ > 7)
```

Reply:
Note that this does not compile:  https://godbolt.org/z/E5c3v6

I'm not sure the Standard guarantees that a compiler won't warn that __GNUC__ is not defined when it compared to 7.

Reply:
hmm, we have a version check of `__GNUC__` in InkAPITest.cc. Is this broken?
https://github.com/apache/trafficserver/blob/89b6b91d43eb43d16038ef8e9f9a9a1cfd2ed581/src/traffic_server/InkAPITest.cc#L25

Reply:
This looks standard approach
```
/* Test for GCC > 3.2.0 */
#if __GNUC__ > 3 || \
    (__GNUC__ == 3 && (__GNUC_MINOR__ > 2 || \
                       (__GNUC_MINOR__ == 2 && \
                        __GNUC_PATCHLEVEL__ > 0))
```
https://gcc.gnu.org/onlinedocs/cpp/Common-Predefined-Macros.html

Reply:
Oh, I got point we need `static_assert` if `__GNUC__` is not defined for clang.

Reply:
Another approach to avoid these preprocessors is adding a message as the second argument to use c++11 version of `static_assert(bool_constexpr , message)` .

> message can be omitted. (since C++17)

Reply:
That doesn't seem to work:  https://godbolt.org/z/17cfTo

Reply:
`#if defined(__GNUC__) && (__GNUC__ >= 4) && (__GNUC_MINOR__ >= 3)` does work for clang:        https://godbolt.org/z/bGEz9P

But I'm not sure the Standard guarantees it will work for a different compiler that we may want to use in the future.

Reply:
Talked on slack, this is a workaround fix to avoid a bug of gcc-7 with static_assert. Let's remove when we drop the gcc-7 support.